### PR TITLE
Add traceback to unhandled promise rejections, Fixes: #14631

### PR DESCRIPTION
**Pull request by graingert**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Feature

**What is the current behavior? (You can also link to an open issue here)**
Console is filled with meaningless:  `Possibly unhandled rejection: {}` errors

**What is the new behavior (if this is a feature change)?**
Console is filled with juicy traceback info.

**Does this PR introduce a breaking change?**
I don't think so

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

### CODE REVIEW

**Comment by graingert** [comment]
@gkalpak uhh, are the tests for the unhandled rejection even run? I added these to intentionally fail...

**Comment by gkalpak** [comment]
Not sure what you mean. The tests should run (and I am quite sure they do).
(The tests need some fixes, because they are not currently testing what they should be testing, but I that is unrelated.)

**Comment by graingert** [comment]
@gkalpak https://github.com/angular/angular.js/pull/15527/files#diff-0cfb390955e411a4a5437fa337d52481R2181 should fail on the fixture.type === 'exception' because it doesn't stringify to 'foo'.

**Comment by graingert** [comment]
@gkalpak are you in IRC?

**Comment by gkalpak** [comment]
@graingert, if you want it to fail, then you better [not reject the promise with `foo`] 😛
<sub>(I am rarely (read "never") on IRC these days, but you can find me on gitter.)</sub>

[1]: https://github.com/angular/angular.js/pull/15527/files#diff-0cfb390955e411a4a5437fa337d52481R2179

**Comment by graingert** [comment]
@gkalpak oh I am a silly... Fixing...

**Comment by gkalpak** [comment]
BTW, it would made our life a tiny bit easier if you followed our [commit message guidelines][1] :wink:

[1]: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit

**Comment by graingert** [comment]
@gkalpak will do once I've got the tests working. Also your URL is out of date: you want https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format

File: src/ng/q.js
Comment (gkalpak) [inline src/ng/q.js]: Here is another idea (either for `toDebugString()` or just here):
If `toCheck.value` has `toString()` method that is not the default `Object.prototype.toString()`, then use that for serializing it. Something similar to [stringify][1].

[1]: https://github.com/angular/angular.js/blob/465d1734559ca4a7f4aa24387060f88fcc53ecb1/src/Angular.js#L1995

**Comment by gkalpak** [comment]
Thx! (My URL was fine btw - just intentionally pointing to the supersection of yours 😛)

**Comment by graingert** [comment]
@gkalpak ah it didn't go to a section for me at all.

**Comment by graingert** [comment]
@gkalpak and we're green!

**Comment by gkalpak** [comment]
@graingert, awesome! Now on to that commit message 😛

File: test/ng/qSpec.js
Comment (gkalpak) [inline test/ng/qSpec.js]: Change "exception" to "Error object".

File: test/ng/qSpec.js
Comment (gkalpak) [inline test/ng/qSpec.js]: Instead of repeating `type` in all `it` blocks, create a `describe block for each fixture and put `type` in its description.

File: test/ng/qSpec.js
Comment (gkalpak) [inline test/ng/qSpec.js]: Extra points for putting some space before `it` blocks :smiley:

File: test/ng/qSpec.js
Comment (gkalpak) [inline test/ng/qSpec.js]: A couple more fixtures would be in order imo. E.g.: 

- One with a non-`Error` object.
- One with an object that "extends" `Error`.

**Review by gkalpak — CHANGES_REQUESTED** [review]
I left some comments, but the general idea LGTM.

File: test/ng/qSpec.js
Comment (graingert) [inline test/ng/qSpec.js]: Cool

File: test/ng/qSpec.js
Comment (graingert) [inline test/ng/qSpec.js]: Yeah makes sense

File: test/ng/qSpec.js
Comment (graingert) [inline test/ng/qSpec.js]: Ah makes sense

**Comment by graingert** [comment]
@gkalpak done

File: test/ng/qSpec.js
Comment (gkalpak) [inline test/ng/qSpec.js]: Now everyone is capitalized except for `plain value`. Just because it is nothing more than a plain value, doesn't mean you have to treat it like a 2nd-class citizen 😛

File: test/ng/qSpec.js
Comment (gkalpak) [inline test/ng/qSpec.js]: Or `describe` blocks...

File: test/ng/qSpec.js
Comment (gkalpak) [inline test/ng/qSpec.js]: Nit: Using something like `'with ' + type` will make the description read better (in case of an error for example).

**Review by gkalpak — APPROVED** [review]
A couple of nitpicks, but otherwise LGTM.

File: test/ng/qSpec.js
Comment (graingert) [inline test/ng/qSpec.js]: I went with lower case, and more accurate names

**Comment by graingert** [comment]
@gkalpak done

**Review by gkalpak — APPROVED** [review]
LGTM, but let's wait for some feedback from others as well.
<sub>(In the mean time, feel free to fix up that commit message :grin:)</sub>

**Comment by graingert** [comment]
@gkalpak I did fix the commit message... What's wrong with it?

**Comment by gkalpak** [comment]
Actual:

> fix($q): Add traceback to unhandled promise rejections, Fixes: #14631

Expected something like:

> fix($q): add traceback to unhandled promise rejections
>
> Fixes #14631

**Comment by mgol** [comment]
@graingert Thanks for the PR, this is awesome. 👏 

I don't have any remarks apart from the incorrect commit message format.

**Comment by graingert** [comment]
@gkalpak commit message fixed

File: src/ng/q.js
Comment (mgol) [inline src/ng/q.js]: Is this the correct order? It looks weird when the error message is *after* the stack trace.

**Review by mgol — CHANGES_REQUESTED** [review]

File: src/ng/q.js
Comment (mgol) [inline src/ng/q.js]: Hmm, OK. I mostly wondered because this:
screen shot 2016-12-21 at 15 48 43
looks a little weird.

Comment (mgol) [inline src/ng/q.js]: There is nothing in the `$exceptionHandler` code that would make the order important; that's all of it:
```js
function $ExceptionHandlerProvider() {
  this.$get = ['$log', function($log) {
    return function(exception, cause) {
      $log.error.apply($log, arguments);
    };
  }];
}
```

I wonder what's the reason for the current documentation. @gkalpak, @petebacondarwin does any one of you know?

Comment (graingert) [inline src/ng/q.js]: what about just doing
```js
exceptionHandler(toCheck.value, 'Possibly unhandled rejection')
```

Comment (mgol) [inline src/ng/q.js]: That sounds sensible but I'd still swap the order of operations. Note that currently we are trying to show the `'Possibly unhandled rejection'` message first and the error afterwards; we just put it all in one string argument so the stack trace gets lost. So I'd like:
```js
exceptionHandler('Possibly unhandled rejection', toCheck.value);
```
but that would require updating the docs to no longer require the cause to be the second parameter (no code changes required as it already works as I mentioned before).

I'd just want to know if there's any reason for the current signature so let's see what others say. :)

Comment (graingert) [inline src/ng/q.js]: @mgol my custom $exceptionHandlers are configured to pass the first argument as the Error value, as per the documentation.

I presume others follow the same documentation also.

Comment (graingert) [inline src/ng/q.js]: @mgol swapping the types of the parameters of $exceptionHandler is another issue for another PR, and would be a breaking change.

Comment (mgol) [inline src/ng/q.js]: OK, fair enough. In that case, maybe it's better to leave it as it is for now; your proposal to just write:
```js
exceptionHandler(toCheck.value, 'Possibly unhandled rejection');
```
sounds nice but we pass other data types through `toDebugString` and we'd lose it in that case. If we want to tweak that, it would be better to tackle that separately after this PR lands.

**Review by mgol — APPROVED** [review]

File: src/ng/q.js
Comment (dcherman) [inline src/ng/q.js]: Though I don't think it necessarily a concern here, do we care that this will not identify errors that originate from an iframe?  Don't know how common of a use case that is.

Comment (mgol) [inline src/ng/q.js]: I doubt it's common but if it turns out to be an issue we can revisit.

Ideally we would convert to a string (via `toDebugString`) only certain types of values and log others unchanged. The problem is that then we'd have to pass `Possibly unhandled rejection` and `toCheck.value` as separate arguments to `$exceptionHandler` and if we do it in a documented way (error first, (optional) reason second) the message would get swapped from e.g.:
```
'Possibly unhandled rejection: MY_THING'
```
to:
```
'MY_THING' 'Possibly unhandled rejection'
```
which is less pleasant.

We can discuss it but let's do it in a separate issue if desired.

**Comment by mgol** [comment]
Landed, thanks!

**Comment by mgol** [comment]
@graingert One nit (I fixed it myself): there should be no colon after `Fixes` in the commit description.

**Pull request closed without merging**
